demos/TicTacToe/Classes/main_menu_scene.cc

```diff
@@ -360,64 +364,77 @@ bool MainMenuScene::init() {
-                              cocos2d::Event* event) -> bool {
-    // Returns false, not consuming the event, to exit the layer if
-    // current_state_ is not in the kGameMenuState or is switching states.
-    if (previous_state_ != current_state_ || current_state_ != kGameMenuState) {
-      return false;
-    }
-    const auto bounds = event->getCurrentTarget()->getBoundingBox();
-    const auto point = touch->getLocation();
-    if (bounds.containsPoint(point)) {
-      // Show the on screen keyboard and places character inputs into the text
-      // field.
-      auto str = join_text_field->getString();
-      auto text_field = dynamic_cast<TextFieldTTF*>(event->getCurrentTarget());
-      text_field->setCursorEnabled(true);
-      text_field->attachWithIME();
-    } else {
-      auto text_field = dynamic_cast<TextFieldTTF*>(event->getCurrentTarget());
-      text_field->setCursorEnabled(false);
-      text_field->detachWithIME();
-    }
 
-    return true;
-  };
+  auto join_text_field_position = Size(480, 95);
+  auto join_text_field_size = Size(180, 80);
+  auto join_text_field = ui::TextField::create("code", "Arial", 48);
+  join_text_field->setTextHorizontalAlignment(TextHAlignment::CENTER);
+  join_text_field->setPosition(join_text_field_position);
+  join_text_field->setAnchorPoint(Vec2::ANCHOR_MIDDLE);
+  join_text_field->setTouchSize(join_text_field_size);
+  join_text_field->setTouchAreaEnabled(true);
+  join_text_field->setMaxLength(/*max_characters=*/4);
+  join_text_field->setMaxLengthEnabled(true);
+
+  // Adds the event listener to handle the actions for a textfield.
+  join_text_field->addEventListener(
+      [this](Ref* sender, ui::TextField::EventType type) {
+        auto join_text_field = dynamic_cast<ui::TextField*>(sender);
+        string join_text_field_string = join_text_field->getString();
+        // Transforms the letter casing to uppercase.
+        std::transform(
+            join_text_field_string.begin(), join_text_field_string.end(),
+            join_text_field_string.begin(),
+            [](unsigned char c) -> unsigned char { return std::toupper(c); });
+        // Creates a repeating blink action for the cursor.
+        const auto repeated_blink = CreateBlinkingCursorAction(join_text_field);
+        switch (type) {
+          case ui::TextField::EventType::ATTACH_WITH_IME:
+            // Runs the repeated blinking cursor action.
+            join_text_field->runAction(repeated_blink);
+            CCLOG("displayed keyboard");
+            break;
+          case ui::TextField::EventType::DETACH_WITH_IME:
+            // Stops the blinking cursor.
+            join_text_field->stopAllActions();
+            CCLOG("dismissed keyboard");
+            break;
+          case ui::TextField::EventType::INSERT_TEXT:
+            join_text_field->setString(join_text_field_string);
+            CCLOG("join_text_field inserted text : %s", join_text_field_string);
```

Comment:
Now that your code works, are these logs useful for anything?  Can they be removed?